File: src/main/java/org/danilopianini/util/SpatialIndex.java
package org.danilopianini.util;

import java.io.Serializable;
import java.util.List;

/**
 *
 * @param <E> content of the index
 */
public interface SpatialIndex<E> extends Serializable {

    /**
     * Inserts an element in the {@link SpatialIndex}.
     * 
     * @param element
     *            The element to add
     * @param position
     *            the coordinates where the element should be added
     */
    void insert(E element, double... position);

    /**
     * Deletes an element from the {@link SpatialIndex}.
     * 
     * @param element
     *            The element to add
     * @param position
     *            the coordinates where the element should be added
     * @return true if the element is found and removed
     */
    boolean remove(E element, double... position);

    /**
     * If an element is moved, updates the {@link SpatialIndex} accordingly.
     * 
     * @param element
     *            The element to add
     * @param start
     *            the coordinates where the element is currently located
     * @param end
     *            the coordinates where the element is should be moved
     * @return true if the element is found and moved
     */
    boolean move(E element, double[] start, double[] end);

    /**
     * Queries the {@link SpatialIndex}, searching for elements in a
     * parallelotope. "Parallelotope" is a fancy word for
     * "N-dimensional rectangle". A 2-parallelotope is in fact a rectangle, a
     * 3-parallelotope is a parallelepiped, and so on.
     * 
     * @param parallelotope
     *            the space where to search for elements
     * @return the list of elements in this area of the {@link SpatialIndex}.
     */
    List<E> query(double[]... parallelotope);

    /**
     * @return the number of dimension of space for this {@link SpatialIndex}.
     */
    int getDimensions();

}

File: src/main/kotlin/org/danilopianini/util/SpatialIndex.kt
package org.danilopianini.util

import java.io.Serializable

/**
 * @param E content of the index
 */
interface SpatialIndex<E> : Serializable {
    /**
     * Inserts an element in the [SpatialIndex].
     *
     * @param element The element to add
     * @param position The coordinates where the element should be added
     */
    fun insert(
        element: E,
        vararg position: Double,
    )

    /**
     * Deletes an element from the [SpatialIndex].
     *
     * @param element The element to remove
     * @param position The coordinates where the element should be removed from
     * @return true if the element is found and removed
     */
    fun remove(
        element: E,
        vararg position: Double,
    ): Boolean

    /**
     * If an element is moved, updates the [SpatialIndex] accordingly.
     *
     * @param element The element to move
     * @param start The coordinates where the element is currently located
     * @param end The coordinates where the element should be moved
     * @return true if the element is found and moved
     */
    fun move(
        element: E,
        start: DoubleArray,
        end: DoubleArray,
    ): Boolean

    /**
     * Queries the [SpatialIndex], searching for elements in a parallelotope.
     * "Parallelotope" is a fancy word for "N-dimensional rectangle".
     * A 2-parallelotope is a rectangle, a 3-parallelotope is a parallelepiped, and so on.
     *
     * @param parallelotope The space where to search for elements
     * @return The list of elements in this area of the [SpatialIndex].
     */
    fun query(vararg parallelotope: DoubleArray): List<E>

    /**
     * @return The number of dimensions of space for this [SpatialIndex].
     */
    val dimensions: Int
}
